homeassistant/components/openweathermap/forecast_update_coordinator.py

```diff
@@ -0,0 +1,137 @@
+        """Initialize coordinator."""
+        self._owm_client = owm
+        self._forecast_mode = forecast_mode
+        self._latitude = latitude
+        self._longitude = longitude
+        self._forecast_limit = 15
+
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=FORECAST_UPDATE_INTERVAL
+        )
+
+    async def _async_update_data(self):
+        data = {}
+        with async_timeout.timeout(20):
+            try:
+                forecast_response = await self._get_owm_forecast()
+                data = self._convert_forecast_response(forecast_response)
+            except (APICallError, UnauthorizedError) as error:
+                raise UpdateFailed(error) from error
+
+        return data
+
+    async def _get_owm_forecast(self):
+        if self._forecast_mode == "daily":
+            forecast_response = await self.hass.async_add_executor_job(
+                self._owm_client.daily_forecast_at_coords,
+                self._latitude,
+                self._longitude,
+                self._forecast_limit,
+            )
+        else:
+            forecast_response = await self.hass.async_add_executor_job(
+                self._owm_client.three_hours_forecast_at_coords,
+                self._latitude,
+                self._longitude,
+            )
+        return forecast_response.get_forecast()
+
+    def _convert_forecast_response(self, forecast_response):
+        weathers = self._get_weathers(forecast_response)
+
+        forecast_entries = self._convert_forecast_entries(weathers)
+
+        return {
+            ATTR_API_FORECAST: forecast_entries,
+            ATTR_API_THIS_DAY_FORECAST: forecast_entries[0],
+        }
+
+    def _get_weathers(self, forecast_response):
+        if self._forecast_mode == "freedaily":
+            return forecast_response.get_weathers()[::8]
+        return forecast_response.get_weathers()
+
+    def _convert_forecast_entries(self, entries):
+        if self._forecast_mode == "daily":
+            return list(map(self._convert_daily_forecast, entries))
+        return list(map(self._convert_forecast, entries))
+
+    def _convert_daily_forecast(self, entry):
+        return {
+            ATTR_FORECAST_TIME: entry.get_reference_time("unix") * 1000,
+            ATTR_FORECAST_TEMP: entry.get_temperature("celsius").get("day"),
+            ATTR_FORECAST_TEMP_LOW: entry.get_temperature("celsius").get("night"),
+            ATTR_FORECAST_PRECIPITATION: self._calc_daily_precipitation(
+                entry.get_rain().get("all"), entry.get_snow().get("all")
+            ),
+            ATTR_FORECAST_WIND_SPEED: entry.get_wind().get("speed"),
+            ATTR_FORECAST_WIND_BEARING: entry.get_wind().get("deg"),
+            ATTR_FORECAST_CONDITION: self._get_condition(entry.get_weather_code()),
+        }
+
+    def _convert_forecast(self, entry):
+        return {
+            ATTR_FORECAST_TIME: entry.get_reference_time("unix") * 1000,
```

Comment:
The forecast time should be in UTC datetime isoformat.
https://developers.home-assistant.io/docs/core/entity/weather#forecast